Redesigned the Symfony Profiler

javiereguiluz opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |
### Context

Recently, we redesigned the web debug toolbar ([read announcement](http://symfony.com/blog/new-in-symfony-2-8-redesigned-web-debug-toolbar)). We maintained "the spirit" of the original toolbar, but its visual appearance was completely changed.

Now it's the turn of the Symfony Profiler. We want to update it to match the new toolbar design and to give it a more modern look and feel.
### The redesign

Most of the redesign is focused on updating just the visual appearance, but we've seized this opportunity to make some minor functional changes as well. Testing this pull request in your own applications may be difficult, so we'll provide below a complete walkthrough of the redesign. (Click on the images to see them in full resolution)
### Overview

This is the new general structure of the profiler pages:

![profiler-structure](https://cloud.githubusercontent.com/assets/73419/9226162/c1bf71f0-410d-11e5-8a7b-5c690b7c8c8a.png)
### Request panel
- Renamed `Request` panel to `Request / Response`
- Divided contents into tabs: Request, Response, Session, etc. (this prevents ultra long pages)

![profiler-request](https://cloud.githubusercontent.com/assets/73419/9226172/cb7f7938-410d-11e5-8ffd-cea8264446dd.png)
### Timeline panel
- Renamed `Timeline` panel to `Performance`
- Highlighted the most important metrics
- Added "Peak Memory Usage" to those metrics
- The timeline graph is now easier to read (font size is slightly larger, updated the color palette, made the canvas wider, etc.)
- If there is only one request, the page doesn't distract you with the "Master Request" heading. If there are several sub-requests, headings are now bigger and graphs are more separated.

![profiler-performance](https://cloud.githubusercontent.com/assets/73419/9226178/d1335782-410d-11e5-9f9a-5ace45b0dfa0.png)
### Twig panel
- Just a minor update to highlight the important metrics. The rest of the contents remain the same.

![profiler-twig](https://cloud.githubusercontent.com/assets/73419/9226180/d5bd6ff4-410d-11e5-86c8-beba426516c1.png)
### Exception panel
- Just minor design tweaks.

![profiler-exception](https://cloud.githubusercontent.com/assets/73419/9226182/da0dfd9e-410d-11e5-965c-e980f7144541.png)
### Events panel

When the `framework.ide` config option is set:
- The listeners are now displayed as more evident clickable links.
- Bellow the listener we show its full class namespace (if you don't need this information, it doesn't distract you much; if you need it, you don't have to wait for the `<abbr>` tooltip to show it)

![profiler-events-links-enabled](https://cloud.githubusercontent.com/assets/73419/9226188/e403ae48-410d-11e5-8c86-5d28fed940e7.png)

When the `framework.ide` config option is NOT set:
- In this case the developer cannot click on the listener name to go to that file and line. Therefore, we think it's important to display that information (full class namespace + full file path + line number). At first it may seem that it contains too much information ... but this panel should be more "useful" than "beautiful".

![profiler-events-links-disabled](https://cloud.githubusercontent.com/assets/73419/9226198/eb9f3bae-410d-11e5-8bbe-95c148809387.png)
### Logs panel
- Logs are now grouped in: "Info. & Errors" (this is the panel that you need most of the times); "Deprecations" (separated from the normal logs because they are specific to the process of updating your Symfony version); and "Debug" (which are all those messages that you almost never need)
- We now show the channel for the log messages (for built-in Symfony channels it may be redundant, but your application can define custom channels)

![profiler-logger](https://cloud.githubusercontent.com/assets/73419/9226203/f345163a-410d-11e5-85e6-cf9fceb885fc.png)
### Routing panel
- Just a minor redesign and highlighted some information.

![profiler-routing](https://cloud.githubusercontent.com/assets/73419/9226213/00e99130-410e-11e5-90c0-dda9394dabd7.png)
### Forms panel
- Not redesigned yet.
### Translation panel
- Highlighted the important metrics.
- Replaced the old selector by the new tabbed navigation.

![profiler-translation](https://cloud.githubusercontent.com/assets/73419/9226215/06036c40-410e-11e5-90c9-fe3b43dd6156.png)
### Security panel
- Just a minor redesign and highlighted some information.

![profiler-security](https://cloud.githubusercontent.com/assets/73419/9226217/0a92fbcc-410e-11e5-9438-e9e230a7865b.png)
### E-mails panel
- Not redesigned yet.
### Doctrine panel
- Queries are no longer displayed "shrinked". We display the full query without parameters by default.
- The three old actions ("expand", "runnable" and "explain") have been reduced to two actions ("runnable" and "explain")

![profiler-doctrine](https://cloud.githubusercontent.com/assets/73419/9226225/13e0b5d4-410e-11e5-96ce-1eb67f99df49.png)

Here you can see what happens when you click on the "Show runnable query" and "Explain query" links:

![profiler-doctrine-detail](https://cloud.githubusercontent.com/assets/73419/9226247/33b47918-410e-11e5-9657-e1351787657a.png)
### Debug panel
- Not redesigned yet.
### Config panel
- This was probably the "ugliest" panel in the old profiler, because of the two first tables.
- We now display the information of those tables highlighted in a different way.

![profiler-config](https://cloud.githubusercontent.com/assets/73419/9226257/3d728850-410e-11e5-86d8-da807f6a1217.png)
### Search results

Results with response status different from 2xx now are highlighted:

![profiler-search-results](https://cloud.githubusercontent.com/assets/73419/9226279/44bacd8e-410e-11e5-8daf-e77d2be7e3c6.png)
### Other design elements

The page title now changes for each section:

**Before**

![page-title-before](https://cloud.githubusercontent.com/assets/73419/9226312/6301b9c4-410e-11e5-8ef7-001be222ac53.png)

**After**

![page-title-after](https://cloud.githubusercontent.com/assets/73419/9226316/693bce42-410e-11e5-897a-c5e4413bc638.png)

Empty elements now display a dashed border. This allows to make design more consistent between different section. Besides, once you are used to the new design, your brain will immediately associate that design to an empty element, so you'll no longer need to read the inner text content.

![profiler-empty-element](https://cloud.githubusercontent.com/assets/73419/9226400/c9ffe06a-410e-11e5-9859-f095e444d9af.png)

Besides, when the entire panel doesn't have any content, its menu label looks disabled:

![sidebar-menu](https://cloud.githubusercontent.com/assets/73419/9226405/cf4ee5e8-410e-11e5-9501-523cd6b3413e.png)

Long pages have been break down using a tabs-based navigation:

![tabbed-navigation](https://cloud.githubusercontent.com/assets/73419/9226409/d918dba6-410e-11e5-9cd4-516a082950de.gif)

Below the page header we display a "summary section" which displays the profiled URL and some useful profiling information. The big URL is the natural way to "go back to application" from the profiler:

![summary-back-to-application](https://cloud.githubusercontent.com/assets/73419/9226414/df1abb00-410e-11e5-88ff-59dbd35b584f.gif)

Besides, the summary background color changes according to the response HTTP status code:

![profiler-summary](https://cloud.githubusercontent.com/assets/73419/9226423/e5b03850-410e-11e5-9163-2f2153c5d1de.png)

Above the main menu we now display two shortcuts to common operations: "display the 10 most recent profiles" and "search profiles":

![sidebar-shortcuts](https://cloud.githubusercontent.com/assets/73419/9226427/eaea4586-410e-11e5-8ff4-f6141723996a.png)

The search form in the sidebar has been collapsed and it's revealed when clicking on the "Search" shortcut button:

![search-form-reveal](https://cloud.githubusercontent.com/assets/73419/9226435/ef907fec-410e-11e5-9627-37cb2a241f86.gif)

---

And that's all! Please tell us what you think about this redesign. **Don't hesitate to criticize anything you don't like or think it's wrong or it's worse than in the previous profiler.** Let's work together to make the new profiler a reality! Thanks!

javiereguiluz committed:
Redesigned most of the Symfony Profiler panels

romqin commented:
:clap: :+1:

DZunke commented:
it seems to be far more accessible then the old version. so ... vote :+1: for this. 

great work! :clap:

ogizanagi commented:
Big :+1: :clap: :clap: !
Fantastic job !

deguif commented:
I really like the new design.
But from the screenshots, there's only one thing that appears confused to me with the new design.
In the Exception panel, the line causing the exception seems to not being highlighted any more in the stack trace. Is there any reason ? ping @javiereguiluz

romqin commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/layout.html.twig:
```diff
@@ -1,131 +1,92 @@
-                    menuItems[elem].children.length > 0) {
-                    child = menuItems[elem].children[0];
+                    <label for="reveal-search-form" class="btn btn-sm">
+                        {{ include('@WebProfiler/Icon/search.svg') }} Search
```
This icon is missing ?

javiereguiluz committed:
Added some missing SVG icons

romqin commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -76,193 +143,422 @@ abbr {
+
+.container {
+    min-width: 960px;
+    max-width: 1200px;
```
In your screenshots it renders perfectly, but why limit the width ?

Because on a big screen this look like this:
![capture d ecran 2015-08-12 a 17 21 58](https://cloud.githubusercontent.com/assets/334432/9227977/8ca1de82-4116-11e5-8e29-3523f5d836d9.png)

oyepez003 commented:
+10000

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/layout.html.twig:
```diff
@@ -1,131 +1,92 @@
+            </div>
+        </div>
+    {% elseif profile is defined %}
+        {% set status_code = profile.getcollector('request').statuscode|default(0) %}
```
what if the request collector is not registered ? The template should account for it

javiereguiluz commented:
@romqin I've added the missing SVG icons. Thanks for reporting the error.

@deguif very nice catch! I totally forgot about that. The line is now highlighted again:

![exception_line_highlight](https://cloud.githubusercontent.com/assets/73419/9228394/9211b8ae-4118-11e5-9643-5f081ad232d7.png)

@romainneutron the problem is that at some point, making the lines longer worsen readability instead of improving it. The classic limit in the print era was around 75 characters per line (including white spaces). The longest line in the current table has 96 chars (and the longest file path text contain up to 134 chars, but it's not an important element).

javiereguiluz committed:
Highlight the line of code which caused the exception

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -295,114 +591,220 @@ td.main, td.menu {
+}
+
+
+{% verbatim %}
```
is a verbatim tag actually needed here ? and why only on a small part of the template ?

javiereguiluz replied:
Without the verbatim tag, the selector `#tabs input[type=radio]:checked + label` produces an error.

But don't worry about this. The problem is that I dreamed about creating a "pure CSS" design. The tabs and the search form toggle are implemented just with CSS. However, this requires some ugly CSS tricks so the dream turned into a nightmare. I'm going to update this to use JavaScript.

stof replied:
@javiereguiluz which error ? that look weird as I don't see anything looking like the beginning of a Twig block here. This may be a bug in the Twig regex

javiereguiluz replied:
I've removed the `verbatim` tag and the error no longer appears :astonished:

OskarStark commented:
awesome!!!!

OskarStark commented:
the height of the input and the button is different, isn't it?

![bildschirmfoto 2015-08-12 um 17 57 42](https://cloud.githubusercontent.com/assets/995707/9229095/b9283f6e-411b-11e5-90d2-c7ef76182277.png)

iltar commented:
@javiereguiluz I love the redesign, it's more functional and just looks awesome! One question though, is it by design the code from the exception will remain light themed?

stof commented:
@javiereguiluz can you also open the DoctrineBundle PR so that I can see what it implies for a bundle supporting both versions of the profiler ?

WouterJ commented:
In general: Perfect!

In more detail: I would change some things :) Some thoughts after playing with it for some minutes:

* The design now takes the full width, instead of the previous centered-fixed-width-box. However, the content still has this very small width. This looks very weird on a bigger screen. The most strange thing is the search box floating somewhere in the center of the page, while its look reveals it should be right aligned:

   ![sf-profiler-fixed-width](https://cloud.githubusercontent.com/assets/749025/9230995/ebf08a0a-4125-11e5-8338-d83f2876911e.png)

* I like the color of the summary bar. However, I find it too big. What about merging the 2 lines into one? (or moving most of the information in the, almost empty, top bar?)
* I don't like the new yellow highlight color, it's almost invisible.
* I don't like it that "Symfony" after the Symfony logo in the top bar doesn't use the Symfony logo font, but some Helvetica/Arial.
* Due to how a human eye handles colors, selected tabs look less big (as the fill color is darker than the border color) than not selected tabs. This doesn't feel very natural. I propose to make the border darker (`#555`) when a tab is selected.
* In the menu, when hovering over menu items, the current menu items looses its focus. I like it more if the current menu item is always highlighted.

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -76,193 +143,422 @@ abbr {
+    margin-bottom: 0;
+}
+#menu-profiler li:last-child {
+    bborder: 0;
```
typo?

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -295,114 +591,220 @@ td.main, td.menu {
-    width: 160px;
+#tabs input[type=radio]:checked + label {
+    background: #666;
+bbox-shadow: inset 1px 1px 2px #CCC;
```
typo?

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -1,71 +1,138 @@
+    width: 19%;
+}
+table tbody th {
+    bbackground-color: #FFF;
```
typo?

javiereguiluz committed:
CSS tweaks to fix some reported issues

ogizanagi commented on src/Symfony/Bundle/DebugBundle/Resources/views/Profiler/dump.html.twig:
```diff
@@ -39,21 +36,19 @@
-        <span class="count">
-            <span>{{ collector.dumpsCount }}</span>
-        </span>
+    <span class="label {{ collector.dumpsCount ? '' : 'disabled' }}">
```
`{{ not collector.dumpsCount ? 'disabled' }}`

ogizanagi replied:
`{{ collector.dumpsCount == 0 '' ? 'disabled' }}`

javiereguiluz commented:
@WouterJ thanks for your review. I've made some tweaks:

  * The selected menu item remains selected when "hovering" the mouse.
  * The selected navigation tab border color matches its background to aovid the 1px issue.
  * The shade of the `.status-warning` yellow is slightly darker
  * The "Symfony Profiler" title font will be tweaked once our designer returns from his summer holidays :)

Regarding the "summary section", for now we prefer to maintain it as it is. first, because this colored block balances the "too gray / too dark" feeling of the rest of the page. Second because the green/yellow/red background is a very useful visual clue (assuming the user doesn't have problems perceiving colors). Third, the URL looks very short in your example ... but most of the times is way longer (and sometimes very long and it spans two lines!)

Lastly, regarding the behavior with high resolution screens, here it is the comparison between maintaining it on the left or centering it. Let's do a quick poll to see what do people like:

![wide-screen-left](https://cloud.githubusercontent.com/assets/73419/9235198/7e34ba50-413d-11e5-977f-b3b4b656daf3.png)

![wide-screen-center](https://cloud.githubusercontent.com/assets/73419/9235200/8082a6b4-413d-11e5-9714-77d33654bc8d.png)

ogizanagi commented on src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig:
```diff
@@ -46,54 +44,70 @@
-    <span class="icon"><svg width="42" height="30" xmlns="http://www.w3.org/2000/svg" version="1.1" x="0px" y="0px" viewBox="0 0 42 30" enable-background="new 0 0 42 30" xml:space="preserve"><g><path fill="#3F3F3F" d="M16.1 15.3c0-0.5 0-0.8 0-1.2c0.2-0.1 0.6-0.8 0.7-1.4c0.2 0 0.4-0.2 0.5-0.8c0-0.4-0.1-0.5-0.2-0.6 c0.3-0.8 0.8-3.4-1.1-3.7C15.7 7.2 15.2 7 14.6 7c-2.6 0-2.9 2-2.4 4.2c-0.1 0.1-0.3 0. ...
-    <strong>Security</strong>
-</span>
+    <span class="label {{ collector.enabled ? '' : 'disabled' }}">
```
`{{ not collector.enabled ? 'disabled' }}`

WouterJ commented:
I prefer the second screenshot (centered page).

ogizanagi commented:
@javiereguiluz : I prefer the first screenshot, but it's right some of the tables deserve more space.

Regarding the 2 heading bars, what do you think about merging them ?
![7e34ba50-413d-11e5-977f-b3b4b656daf3 copie](https://cloud.githubusercontent.com/assets/2211145/9235589/e6db28ee-413f-11e5-939d-9a1d96ba4da0.png)
 And eventually making it smaller ?

hugohenrique commented:
awesome! :+1:

mvhirsch commented:
awesome, please add! :+1:

OskarStark commented:
:+1:  for centered

javiereguiluz commented:
@WouterJ @OskarStark the problem is that the layout wasn't designed to be displayed centered and it looks really bad :cry:

OskarStark commented:
hmm, ok then i would leave it as it is und maybe fix the whole problem to center the page later in a separate PR

apfelbox commented:
I really love the new design, just a few smaller remarks:

*   How often is the Search on Symfony website” actually used?
*   We have a lot of “code” content, so probably we should use a monotype font more often? Practically every key-value table (like in the request tab) or especially things like the paths in the Routing panel or the context of the log messages.
*   I love the color coding of the header! It might be difficult for users with visual impairment, so we could imitate the HTTP Method head? So the line where currently just the URL is, could look like `POST 500 http://...`? Just a thought...
*   I like the tabbed content.
*   As @iltar already mentioned: I would probably choose a dark theme for the code lines in the exception panel.
*   Maybe a pretty-JSON formatter for the log context? We could make it expandable, like the doctrine queries are in the current (2.7) version.
*   The shade of green in the routing panel is very desaturized
*   Maybe we should show the Roles in the security panel as a list instead of inline JSON.
*   Search results: I love it
*   Empty elements with the dashed border: I also like it, although it looks a little bit like a drag & drop upload field.

iltar commented:
@apfelbox if it's the same search as the website, it will be "offline": http://symfony.com/search?q=form

Koc commented:
@javiereguiluz maybe some commits from https://github.com/symfony/symfony/pull/12363 can be cherry-picked into this PR as improvements of exception page, what do you think?

stof commented:
@Koc I would rather keep it separate to make the review easier

apfelbox commented:
@stof separate PR makes sense, but probably we should adapt the new look & feel for error and exception pages? What do you think @javiereguiluz ?

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -1,71 +1,138 @@
+               Please, don't review its contents yet.
+ */
+
+/*! Bootstrap normalize.css v3.0.3 | MIT License */
```
normalize.css does not come from Bootstrap. It is a separate project (which is indeed used in Bootstrap): https://github.com/necolas/normalize.css

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -76,193 +143,420 @@ abbr {
+    margin-bottom: 10px;
+}
+
+#sidebar #sidebar-shortcuts {
```
why nesting 3 id selector ? this does not make sense

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -76,193 +143,420 @@ abbr {
+#sidebar #sidebar-shortcuts {
+    background: #333;
+    padding: 16px 10px;
+    /* needed to prevent visual issues dut to the pure CSS slideToggle() */
```
typo: ``due``

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -295,114 +589,220 @@ td.main, td.menu {
+.sql-actions {
+    margin-top: 4px;
 }
+.sql-actions a {
```
should it really be called ``sql-actions`` ? What makes it specific to SQL ? Other panels may want to have actions too

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig:
```diff
@@ -295,114 +589,220 @@ td.main, td.menu {
+.sql-error     { color: #CC0000; }
+*/
+
+.sql-keyword   { color: #AA2063; font-weight: bold; }
```
IMO, highlighting CSS rules should not be specific to SQL. what if another panel is highlighting something else than SQL ?

stof commented:
@apfelbox doing it separately also means we should finish this one before doing the other one IMO to avoid a mess (conflicts between them, and very hard to follow the review). And it is out of the scope of the discussion on this PR

javiereguiluz committed:
A new round of tweaks and fixes

javiereguiluz commented:
@apfelbox thanks for your nice and detailed review! I've made a lot of tweaks in a7ae8aa and I've added some ideas to the "Wish List" created in #15439

derrabus commented:
Awesome!

javiereguiluz committed:
Redesigned the "debug/dump" panel and other minor tweaks

javiereguiluz commented:
The "Dump / Debug" panel has just been redesigned. It only contains some minor visual tweaks to match the new style:

![profiler-dump](https://cloud.githubusercontent.com/assets/73419/9253861/8d05cc88-41e0-11e5-8853-66ea1a170537.png)

When you click on the "Show code" link:

![profiler-dump-expanded](https://cloud.githubusercontent.com/assets/73419/9253867/97061e40-41e0-11e5-967b-8c4894433530.png)

WouterJ commented:
:-1: on changing the colors of the code block highlighting. These colors are the colors used for code examples throughout the Symfony docs (and after the exception page, throughout Symfony). We should keep that for consistency. Furthermore, for me the new colors look like there's a layer of dust on top of the code blocks.

sstok commented:
"Furthermore, for me the new colors look like there's a layer of dust on top of the code blocks." LOL :joy:  But I do agree that the colors look consistent.

javiereguiluz committed:
Replaced the CSS-based tabs by proper JavaScript-based tabs

mickaelandrieu commented:
This is so gorgeous <3 great contribution @javiereguiluz!

javiereguiluz committed:
First version of the redesigned Form panel

javiereguiluz committed:
Added the new icons, made the sidebar narrower, added badges in the
sidebar and other minor tweaks

javiereguiluz committed:
More design tweaks

javiereguiluz committed:
Refactored the "summary" section and redesigned all the profiler error pages

javiereguiluz committed:
Refactored the JavaScript code that creates the tabs

javiereguiluz committed:
Added a "toggle" utility

javiereguiluz committed:
Refactored the JavaScript code, used toggles in the Dump panel and
added a new "Card" element

javiereguiluz commented:
A quick update to tell you that we've just finished the redesigning of all the built-in profiler panels. The last one was the "Email" panel:

![profiler-email-panel](https://cloud.githubusercontent.com/assets/73419/9429574/41bd4ba0-49d5-11e5-9dbe-2d46f37818cd.png)

The panel shows/hides some of its headings and information depending on the number of mailers and sent messages:

1) Usual scenario (1 mailer and 1 mail):

![email-profiler-1-mailer-1-message](https://cloud.githubusercontent.com/assets/73419/9429576/49d7d062-49d5-11e5-8591-99d37510606e.png)

2) 1 mailer and several messages:

![email-profiler-1-mailer-2-messages](https://cloud.githubusercontent.com/assets/73419/9429578/4fefb258-49d5-11e5-9897-55dcc99d4403.png)

3) Several mailers and messages:

![profiler-2-mailer-2-messages](https://cloud.githubusercontent.com/assets/73419/9429580/5563d318-49d5-11e5-801f-e17cdddd3f0d.png)

WouterJ commented:
Good job!

Very minor detail: I think there is too much spacing between "E-mails" and "Messages sent by each mailer" in screeshot (3).

Tobion commented:
This is so much better than before. Good job!

2 minor suggestions:

- Events panel without `framework.ide` could highlight the class name without namespace instead of the method name. The method name is kinda the same most of the time ´on...`. The class would be more interesting.
- The headlines capitalization should be unified: E.g. in config panel it's "Symfony Configuration" but also "PHP __c__onfiguration". In Doctrine it's "Database queries" and "Datebase Connections".

WouterJ commented:
I propose using the same title case as used in the docs, @javiereguiluz feel free to ping me if you are unsure what should be capitialized :)

javiereguiluz committed:
Finished the redesign of the Symfony Profiler

javiereguiluz commented:
The time has come :)  I consider the profiler redesign **finished**, so you can now make the final review. I've implemented most, but not all, of the suggestions made so far in the comments of this pull request. Thanks.

mickaelandrieu commented:
please merge! I want to enjoy it ASAP :p

Tobion commented:
:+1: 

Status: Reviewed

stof commented:
Before merging this, I want the upgrade path to be considered for bundles having to be compatible with both designs (which is a must have for high-quality bundles as 2.7 is still supported for 3 years).
The current way used by @javiereguiluz to upgrade bundles is just not maintainable.
See https://github.com/doctrine/DoctrineBundle/pull/452#discussion_r36960481 for the discussion.

stof commented on src/Symfony/Bundle/DebugBundle/Resources/views/Profiler/dump.html.twig:
```diff
@@ -39,42 +35,23 @@
-            {% for dump in collector.getDumps('html') %}
-            <li class="sf-dump sf-reset">
-                in
+    {% if collector.dumpsCount == 0 %}
```
this can be simplified using the ``{% for %} {% else %} {% endfor %}`` Twig construct (it was not possible previously because of the ``ul`` tag)

javiereguiluz replied:
Nice! Thanks.

stof commented on src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig:
```diff
@@ -46,54 +46,69 @@
+                        {{ collector.roles is empty ? 'none' : collector.roles|yaml_encode }}
+
+                        {% if not collector.authenticated and collector.roles is empty %}
+                            <p class="help">User is not authenticated probably because he/she has no roles.</p>
```
``they have`` may be used rather than ``he/she``

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/events.html.twig:
```diff
@@ -1,79 +1,102 @@
 
-    {% if collector.notcalledlisteners %}
-        <table>
+{% macro render_table(events, show_order = false) %}
```
the variable name is weird. This is not a list of events

javiereguiluz replied:
Nice catch! I've replaced it by `listeners`

javiereguiluz committed:
Tweaks as per Stof's review and comments

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig:
```diff
@@ -32,188 +28,179 @@
 {% block menu %}
-    <span class="label">
-        <span class="icon"><svg width="20" height="27" xmlns="http://www.w3.org/2000/svg" version="1.1" x="0px" y="0px" viewBox="0 0 20 27" enable-background="new 0 0 20 27" xml:space="preserve"><g><polygon fill="#3F3F3F" points="16,2 13,2 13,0 7,0 7,2 4,2 4,6 16,6"/><path fill="#3F3F3F" d="M17 3v4H3V3H0v24h20V3H17z M3 13h10v1H3V13z M8 24H3v-1h5V24z M17 22H3v-1h14V22z M17 20H3v-1h14V20z M17 18H3v ...
+    <span class="label {{ collector.data.forms is empty ? 'disabled' }}">
```
shoudln't it also be red when there are errors, to be consistent with the toolbar ?

javiereguiluz replied:
I agree. If there are errors, let's show them prominently. Thanks.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig:
```diff
@@ -530,37 +523,44 @@
+                </thead>
+                <tbody>
+                    <tr>
+                        <th class="font-normal">Model Format</th>
```
shouldn't you add ``scope="row"`` like in other tables you are using ?

javiereguiluz replied:
Yes. Added. Thanks.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig:
```diff
@@ -646,23 +660,26 @@
                 {% for option, value in data.resolved_options %}
                 <tr>
                     <th>{{ option }}</th>
-                    <td><pre>{{ value }}</pre></td>
```
is it intended to remove it ?

javiereguiluz replied:
Yes, because by default we treat all table cells as `<pre>` elements (but breaking long lines to prevent issues).

javiereguiluz committed:
Minor tweaks for the Form panel

javiereguiluz committed:
Allow a toggle to define if its related content is displayed by default
This allows us to display the search form in the error pages and
the search result pages, which improves user experience.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig:
```diff
@@ -33,103 +33,136 @@
     {% if is_deprecation %}
         {% set stack = log.context.stack|default([]) %}
-        {% set id = 'sf-call-stack-' ~ log_index %}
+        {% set id = 'sf-call-stack-' ~ random(999) %}
```
please don't use random(999) to generate unique ids. This is very likely to miss uniqueness.

You should use the index in the loop, plus a prefix different for each loop

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig:
```diff
@@ -33,103 +33,136 @@
 {% endblock %}
 
+{% macro render_table(logs, show_level = false, is_deprecation = false) %}
+    {% import _self as helper %}
```
I would prefer to keep using the ``{% from %}`` tag to import the needed macro, as done previously

javiereguiluz replied:
I've made all the changes suggested in your other comments, but I'm struggling with this one.

First I don't really know why `{% from ... %}` is recommended instead of `{% import ...`. Second, I though that perfixing all utility macros with the `helper.` prefix would be better for code legibility. But of course this could be a bad idea :)  I just need some more comments about this. Thanks.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig:
```diff
@@ -33,103 +33,136 @@
+                        <span class="text-muted nowrap newline">{{ log.timestamp|date('H:i:s') }}</span>
+                    </td>
+
+                    <td class="font-normal text-small text-bold nowrap">{{ log.channel }}</td>
```
``log.channel`` may be undefined if you use another PSR-3 logger than Monolog (it is a MonologBundle concept). Please keep the profiler panel compatible with such case (it is in 2.7)

stof replied:
the only mandatory keys in the DebugLoggerInterface of Symfony are ``timestamp``, ``message``, ``priority`` and ``priorityName``. All others must be checked for existence

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/memory.html.twig:
```diff
@@ -16,9 +16,9 @@
         <div class="sf-toolbar-info-piece">
             <b>PHP memory limit</b>
-            <span>{{ collector.memoryLimit == -1 ? '&infin;' : '%.0f'|format(collector.memoryLimit / 1024 / 1024)|escape }} MB</span>
+            <span>{{ collector.memoryLimit == -1 ? 'Unlimited' : '%.0f MB'|format(collector.memoryLimit / 1024 / 1024)|escape }}</span>
```
you can remove the explicit escaping now, given that the raw string on the left does not contain HTML entities anymore and so the autoescaping would not break the output anymore when escaping the whole ternary statement

javiereguiluz committed:
Replaced a randomly-generated ID by the real log index

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig:
```diff
@@ -31,148 +29,198 @@
+                {% endif %}
+
+                <h3>Request Headers</h3>
+                {{ include('@WebProfiler/Profiler/bag.html.twig', { bag: collector.requestheaders }, with_context = false) }}
```
shouldn't you change the label to ``Header`` like for the response ?

dewos commented:
That's something I've always wondered: why singular "Exception" and "Translation" ... instead of the plural form? (note: some items are fine in singular, for example Doctrine, Twig etc...).

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/time.html.twig:
```diff
@@ -41,63 +41,82 @@
-    <h2>Timeline</h2>
-    {% if collector.events|length %}
-        {{ block('panelContent') }}
+    {% if collector.events is empty %}
```
the memory metrics could still be available in this case though

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/time.html.twig:
```diff
@@ -41,63 +41,82 @@
-        </table>
+        <label for="threshold">Threshold</label>
+        <input type="number" size="3" name="threshold" value="3" min="0"> ms
+        <span class="help">(timeline only displays events slower than this threshold)</span>
```
events are not slower than a threshold. they are longer than a threshold

Tobion replied:
slower = longer. You probably read "lower" in your mind.

WouterJ replied:
There is a difference between slower and longer: If an event is *slower* than another one, it takes *longer* until it is finished.

In this case, I think slower and longer can both be used, but longer seems to be better (not a native though, /cc @weaverryan )

stof replied:
@WouterJ to me, the event is not slow. the code executing during this event is slow, making the event longer (timeline events are just durations)

javiereguiluz replied:
I agree with both of you. I've changed it by:

```
(timeline only displays events with a duration longer than this threshold)
```

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/time.html.twig:
```diff
@@ -41,63 +41,82 @@
-                <td><input type="number" size="3" name="threshold" value="1" min="0"> ms</td>
-            </tr>
-        </table>
+        <label for="threshold">Threshold</label>
```
``for`` references an id. Your input has no id.

javiereguiluz commented:
@dewos once this PR is merged, people will easily test it and hopefully they'll report a lot of minor errors and inconsistencies like the one you mentioned. Merging this is only the first step into creating a polished profiler for 2.8 / 3.0 :)

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig:
```diff
@@ -37,124 +37,161 @@
+        {% elseif message.state == constant('Symfony\\Component\\Translation\\DataCollectorTranslator::MESSAGE_EQUALS_FALLBACK') %}
+            {% set messages_fallback = messages_fallback|merge([message]) %}
+        {% endif %}
+    {% endfor %}
```
adding methods doing this grouping in the DataCollector would allow to make it better than in Twig (which is not well suited for manipulating data to group it)

javiereguiluz replied:
You are right, but I'd prefer to optimize this in the future.

javiereguiluz committed:
Lots of minor tweaks

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig:
```diff
@@ -37,124 +37,161 @@
+            {% set messages_missing = messages_missing|merge([message]) %}
+        {% elseif message.state == constant('Symfony\\Component\\Translation\\DataCollectorTranslator::MESSAGE_EQUALS_FALLBACK') %}
+            {% set messages_fallback = messages_fallback|merge([message]) %}
+        {% endif %}
```
if we add a new state in the future, this would just drop it into nowhere, making it hard to debug why they don't appear (and maybe not making it clear that something is missing btw)

javiereguiluz replied:
You are right again :)  But if we add a new state, I'm afraid that it's not going to be displayed in any way. Sorting messages is only part of the solution. The other part is actually displaying those messages and this is not ready for other states. But we can easily deal with this when the time comes.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -1,4 +1,7 @@
 <script>/*<![CDATA[*/
+    /* Caution: the contents of this file are processed by Twig before loading
+                them as JavaScript source code. Always use /* comments instead
+                of one-line comments to avoid impossible-to-debug side-effects */
```
this could be a Twig comment

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -46,22 +49,36 @@
-            hasClass = function(el, klass) {
-                return el.className && el.className.match(new RegExp('\\b' + klass + '\\b'));
+            hasClass = function(el, cssClass) {
+                if (el.classList) {
```
checking for classLiast support could be done before defining the function instead of doing it on each call of the function. It would be much more efficient

javiereguiluz replied:
I'd prefer to maintain it for code readability reasons.

WouterJ replied:
```
classListIsSupported = "classList" in document.createElement("_");
if (classListIsSupported) {
    hasClass = function (el, cssClass) { return el.classList.contains(cssClass); }
} else {
    hasClass = function (el, cssClass) { return el.className.match(new RegExp('\\b' + cssClass + '\\b')); }
}
```

Is just as readable and more efficient....

javiereguiluz replied:
I can't use the given code (*Uncaught SyntaxError: Unexpected token if*). Please note that these methods are in fact properties of `Sfjs` object.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -46,22 +49,36 @@
+            addClass = function(el, cssClass) {
+                if (!hasClass(el, cssClass)) {
+                    if (el.classList) {
+                        el.classList.add(cssClass);
```
when using classList-based implementation, you don't even need the ``hasClass`` check as classList already takes care of deduplication

javiereguiluz replied:
In this case I also think the current code is more readable.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -46,22 +49,36 @@
             },
 
+            toggleClass = function(el, cssClass) {
+                hasClass(el, cssClass) ? removeClass(el, cssClass) : addClass(el, cssClass);
```
you can use ``el.classList.toggle`` when available

javiereguiluz replied:
And I'd like to maintain this too for readability reasons.

WouterJ replied:
I fail to see why calling a **class**List.**toggle** function inside a function called **toggleClass** would not be readable? (besides being much more efficient)

javiereguiluz replied:
This is the current code:

```js
toggleClass = function(el, cssClass) {
    hasClass(el, cssClass) ? removeClass(el, cssClass) : addClass(el, cssClass);
}
```

And if I understand you right, you are asking me to do this:

```js
toggleClass = function(el, cssClass) {
    if (el.classList) {
        el.classList.toggle(cssClass);
    } else {
        hasClass(el, cssClass) ?  removeClass(el, cssClass) } : addClass(el, cssClass);
    }
}
```

Maybe I'm missing something :confused:

javiereguiluz replied:
In 988f187 I've changed the JavaScript code related to managing CSS classes. Thanks for proposing this improvements.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,112 @@
+
+                        var tabNavigationItem = document.createElement('li');
+                        tabNavigationItem.setAttribute('data-tab-id', tabId);
+                        tabNavigationItem.setAttribute('class', j == 0 ? 'active' : '');
```
I would use ``.className`` here instead

javiereguiluz replied:
Here we are manipulating the `class` attribute of the HTML element. If we change it for `className` I'm afraid that the code no longer works.

WouterJ replied:
`tabNavigationItem.className = ...` is the official JavaScript way of doing this. You should not use setAttribute here.

javiereguiluz replied:
OK. Now I understand the previous comment. Changed. Thanks!

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,112 @@
+                        tabId = tabNavigation[j].getAttribute('data-tab-id');
+                        document.getElementById(tabId).querySelector('.tab-title').style.display = 'none';
+
+                        if (tabNavigation[j].classList.contains('active')) {
```
please use ``hasClass()`` here as the browser may not support classList

javiereguiluz replied:
Very nice catch! Fixed. Thanks.

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,117 @@
+                        var tabTitle = tabs[j].querySelector('.tab-title').innerHTML;
+
+                        var tabNavigationItem = document.createElement('li');
+                        tabNavigationItem.setAttribute('data-tab-id', tabId);
```
please use dataList instead

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,117 @@
+
+                            /* needed because when the tab contains HTML contents, user can click */
+                            /* on any of those elements instead of their parent '<li>' element */
+                            while (activeTab.tagName.toLowerCase() != 'li') {
```
please use strict comparison: !==

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,117 @@
+                                tabNavigation[k].classList.remove('active');
+                            }
+
+                            activeTab.classList.add('active');
```
use `addClass` here

javiereguiluz committed:
Minor JavaScript changes

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,117 @@
+                            }
+
+                            activeTab.classList.add('active');
+                            var activeTabId = activeTab.getAttribute('data-tab-id');
```
use `dataList` here

javiereguiluz replied:
I'm afraid I don't understand this comment. Can you please show me the new code I should use instead of `var activeTabId = activeTab.getAttribute('data-tab-id');`? Thanks!

WouterJ replied:
`var activeTabId = activeTab.dataList.get('tab-id');`

However, it turns out IE 9 doesn't support dataList. So this is something to change when using jQuery (which will become `activeTab.data('tab-id')`.

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,117 @@
+
+                            activeTab.classList.add('active');
+                            var activeTabId = activeTab.getAttribute('data-tab-id');
+                            document.getElementById(activeTabId).style.display = 'block';
```
can't we put this in a CSS class?

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +317,117 @@
+                        var currentContent = toggle.innerHTML;
+                        var originalContent = toggle.getAttribute('data-toggle-original-content');
+                        var altContent = toggle.getAttribute('data-toggle-alt-content');
+                        toggle.innerHTML = currentContent != altContent ? altContent : originalContent;
```
use strict comparison

WouterJ commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base.html.twig:
```diff
@@ -3,21 +3,20 @@
-        </style>
     </head>
-    <body>
+    <body id="{% block body_id %}{% endblock %}">
```
why not using `{% block body_id '' %}`, like below?

WouterJ commented:
Please note https://github.com/symfony/symfony/pull/15523#issuecomment-130742053 , which seems to be completely ignored.

WouterJ commented:
The dump toolbar block showns the number of dumps on that page, which is nice. I think this should be done in the menu item as well.

javiereguiluz commented:
> The dump toolbar block showns the number of dumps on that page, which is nice. I think this should be done in the menu item as well.

That's correct. Why did we do this? In the old profiler, all panels are always enabled. In the new profiler, when some panel contains no information (for example when you dump nothing), that panel is disabled. So, if "dump" is disabled, don't click on it; if it's enabled, you can be certain that it contains dumped contents. In the new profiler, the badges of the menu are reserved to show the number of errors.

javiereguiluz commented:
> :-1: on changing the colors of the code block highlighting. These colors are the colors used for code examples throughout the Symfony docs (and after the exception page, throughout Symfony). We should keep that for consistency.

The colors are exactly the same ... we've only changed their shades. In this case, the consistency should not be achieved with the Symfony docs, but with the rest of the new Symfony profiler design. I prefer to maintain this change, merge it and let people test it before 2.8/3.0 release. If they think it's horrible, be sure they will loudly say so and we'll fix it.

javiereguiluz committed:
Fixed the title of all panels and fixed some JavaScript issues

norkunas commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +318,117 @@
+                                activeTab = activeTab.parentNode;
+                            }
+
+                            for (var k = 0; k < tabNavigation.length; k++) {
```
I think this would be better:
```javascript
for (var k = 0, l = tabNavigation.length; k < l; k++) {
```

javiereguiluz replied:
I think the new code is less readable :(  By the way, a lot of comments are related to JavaScript performance. But please, let's focus a bit. This is not a full-screen 3-D game rendered with JavaScript + canvas that must run at 60 fps. The JavaScript code involved in each profiler page is negligible.

WouterJ commented:
@javiereguiluz thanks for your comments, it turns out I was testing with the old Swiftmailer profiler, resulting in a number disabled in the menu and the other things I mentioned in the Swiftmailer PR. You can ignore all these comments, as the new profiler looks great!

About JS, why don't we actually use jQuery? Given that this is an isolated page and that we are actually recreating jQuery in the Sfjs object...

javiereguiluz commented:
@WouterJ I completely agree with using jQuery (and by the way, Bootstrap too). But let's leave that for a future PR.

javiereguiluz committed:
Refactored some JavaScript code

javiereguiluz committed:
Improved the vertical spacing of tabbed navigation

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -1,7 +1,25 @@
     Sfjs = (function() {
         "use strict";
 
+        var classListIsSupported = 'classList' in document.createElement('_');
```
creating a div would be better than creating a _ element IMO (which could do weird things)

WouterJ replied:
Hmm, let's change it to `'classList' in document.documentElement` (which is what's used by Modernizr). I took `_` from the polyfill in the MDN docs.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -298,7 +302,117 @@
+                }
+
+                /* display the active tab and add the 'click' event listeners */
+                for (i = 0; i < tabGroups.length; i++) {
```
is it actually needed to perform 2 loops over tab groups, with the second loop querying for elements elements inserted in the first loop (and for elements already used in the first loop too) ? couldn't we loop only once ?

javiereguiluz replied:
I tried to avoid a collosal for loop. But we can optimize/improve this in future pull requests once we merge this. Thanks.

javiereguiluz committed:
Minor JavaScript change

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base.html.twig:
```diff
@@ -3,21 +3,23 @@
+            {% block title %}
+                {{ app.request.query.has('panel') ? app.request.query.get('panel')|title|replace({
+                    'Request': 'Request/Response', 'Time': 'Performance', 'Logger': 'Logs', 'Router': 'Routing',
+                    'Form': 'Forms', 'Swiftmailer': 'E-mails', 'Db': 'Doctrine', 'Dump': 'Debug'
```
instead of hardcoding these replacements here, it would be better to provide an easy way for panels to customize their name in the title.
This would keep the name changes in each panel rather than in the layout, and would allow third-party bundles to benefit from the feature too, and not only Swiftmailer and Doctrine which have hardcoded support here.

javiereguiluz replied:
I agree. Panels can now use the following to set their page titles:

```twig
{% block page_title '...' %}
```

And we show it as follows:

```twig
{% block title %}
    {{ block('page_title')|default(app.request.query.get('panel', '')|title) }} (Symfony Profiler)
{% endblock %}
```

javiereguiluz committed:
More CSS tweaks to polish profiler panels

javiereguiluz committed:
Redesigned the search results page

javiereguiluz committed:
Improved the timeline graphics and allowed panels to define their page titles

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base.html.twig:
```diff
@@ -3,21 +3,20 @@
-        </style>
+        <title>
+            {% block title %}
+                {{ block('page_title')|default(app.request.query.get('panel', '')|title) }} (Symfony Profiler)
```
I think this will cause double escaping of the page_title (once inside the block, and once here escaping the output of the block)

javiereguiluz replied:
I've made several tests and it seems to work as expected:

```twig
{% block page_title 'Request & Response' %}
```

![title_1](https://cloud.githubusercontent.com/assets/73419/9544492/9ac26c3a-4d81-11e5-9577-cc212815e4a9.png)

```twig
{% block page_title 'Request & "Response"' %}
```

![title_2](https://cloud.githubusercontent.com/assets/73419/9544500/ac780124-4d81-11e5-92a5-083626d4841d.png)

```twig
{% block page_title 'Request <em>and</em> Response' %}
```

![title_3](https://cloud.githubusercontent.com/assets/73419/9544504/b4a786d0-4d81-11e5-99cc-293043be6773.png)

stof replied:
the simpler implementation would simply to make ``app.request.query.get('panel', '')|title`` the default content of the block

stof replied:
oh, and the block does not seem to be defined in this template

stof replied:
hmm, this is because your block is a constant string, not a variable, and so it is safe for auto-escaping. So it works currently only because of the double auto-escaping, but would break when using variables in the page_title.

javiereguiluz replied:
What would you think if we keep `Symfony Profiler` as the title of any profiler page? It would simplify things a lot and a customized title doesn't provide much value. In fact, a customized title could make the profiler pages harder to find when you have lots of tabs opened.

stof replied:
totally fine with me

hason commented:
The count box in sidebar is on the new line (Firefox):
![screenshot-localhost 8000 2015-08-28 13-50-50](https://cloud.githubusercontent.com/assets/288535/9545616/dcc6437c-4d8b-11e5-81c8-8e93a3e09241.png)

javiereguiluz committed:
Simplified the management of the profiler panel titles

javiereguiluz committed:
Added some metrics about sub-requests in the performance panel

hhamon commented:
Thanks @javiereguiluz! This new redesign WP is really great!

So much wow! :D

ahilles107 commented:
I see seme render issues on Doctrine panel.

![zrzut ekranu 2015-09-01 09 48 24](https://cloud.githubusercontent.com/assets/298610/9598790/ba22538c-508e-11e5-9339-3e647cc5446e.png)

javiereguiluz committed:
Fixed some CSS issues in Firefox

stof commented:
@ahilles107 there is an opened PR on DoctrineBundle to update it for the new design.

@javiereguiluz this makes me think that the new design is not friendly to bundles which have not updated their panels yet.

javiereguiluz commented:
@hason thanks for reporting this error. In ed53b7e I've made soem tweaks for Firefox and it seems to be fixed now:

![firefox_menu_badges](https://cloud.githubusercontent.com/assets/73419/9599341/66284516-5093-11e5-80f9-fb800c570e9d.png)

@ahilles107 it seems that you are not testing the latest version of the Doctrine panel (the "Expand all queries" button is gone for example). Here you can find the latest version: https://github.com/doctrine/DoctrineBundle/pull/452

@stof the problem occurs because of the way I use the `table-layout` property. I'm going to fix it right now.

ahilles107 commented:
@javiereguiluz - yep, it was outdated doctrine bundle - now it looks perfect! Thanks!

javiereguiluz committed:
Allow to disable tabs, tweaked tables layout and other minor tweaks

javiereguiluz committed:
Display the listener priority in the events panel

javiereguiluz commented:
Since this pull request isn't merged yet, I keep adding features. The latest one is to display the priority of the listeners in the event panel (this has been asked for in #11915 and it's also related to #11263).

It looks like it's working, but I confess that I don't know what I'm doing. So please, review carefully the changes in 705bad7. Thanks!

![called_priority](https://cloud.githubusercontent.com/assets/73419/9653415/87f6af84-5223-11e5-9ac4-9d393bcc0008.png)

![not_called_priority](https://cloud.githubusercontent.com/assets/73419/9653417/8a154cee-5223-11e5-826c-6f54a9c75d55.png)

javiereguiluz committed:
Add a check to avoid errors for special pages (exceptions for example)

javiereguiluz committed:
Fixed some test for the TraceableEventDispatcher changes

sstok commented:
Would it be an idea to show the events list with a background color for the first event?

* EventName (bg colored)
* EventName (normal bg color)
* EventName (normal bg color)
* EventName2 (bg colored)

As they are list by name and then sorted by priority, it's easier to see which events are related.

apfelbox commented:
@sstok sort of grouping the called event listeners by event sounds like a good idea. :+1:

javiereguiluz committed:
Sort uncalled events by priority

javiereguiluz commented:
After c8b46e6 changes, uncalled listeners are now displayed sorted by priority:

![uncalled-listeners](https://cloud.githubusercontent.com/assets/73419/9655512/b9c43320-5231-11e5-854d-6fd220f449ec.png)

Could we do something similar for called listeners?

stof commented:
@javiereguiluz ordering by priority first does not make sense: the priority of 2 listeners can be compared only if they are for the same event. Sorting listeners by priority is done per event in the dispatcher.
It would make sense to sort them by event first, and then by priority

javiereguiluz commented:
@stof you are right. I'm going to revert the latest changes. We have 4 options for the events panel:

-----

1) Do nothing else (just display the priority, the event name and the listener).

-----

2) Group listeners by background color (similar to @sstok proposal):

![separated-by-background](https://cloud.githubusercontent.com/assets/73419/9656515/27d62f7e-5239-11e5-81ce-5af4796deaaf.png)

 -----

3) Group listeners by a border line:

![separated-by-lines](https://cloud.githubusercontent.com/assets/73419/9656517/2b214e16-5239-11e5-9266-4971899ac881.png)

-----

4) Group listeners by event name:

![separated-by-event](https://cloud.githubusercontent.com/assets/73419/9656519/2e107f2a-5239-11e5-8750-44b9f4fc971b.png)

ahilles107 commented:
:+1: for option 4

stloyd commented:
Option 4 is most readeable IMO.

ogizanagi commented:
I love the option 4 :+1:

javiereguiluz committed:
Grouped listeners by event in the events panel

javiereguiluz commented:
OK. I've committed the fourth option. Here it is a detailed screenshot in case you want to better analyze it:

![grouped-events-detail](https://cloud.githubusercontent.com/assets/73419/9656925/bb69117c-523c-11e5-8f81-8beb1dabf958.png)

javiereguiluz committed:
Fixed an HTML markup issue for the new events panel

OskarStark commented:
:+1: for option 4

Nicofuma commented:
looks great

javiereguiluz committed:
Adjusted the logger panel after the grouping of related deprecations

javiereguiluz committed:
Tweaked the logger panel stack trace contents

pyrech commented:
The option 4 looks nicer :+1:

javiereguiluz commented:
This PR is now completely finished. We just need some thumbs up from Symfony deciders to merge it. Thanks!

-----

The error reported by Travis seems unrelated:

```
Fatal error: Call to undefined method 
Symfony\Component\DependencyInjection\ContainerBuilder::addScope() in 
/home/travis/build/symfony/symfony/src/Symfony/Bundle/TwigBundle/vendor/symfony/
framework-bundle/FrameworkBundle.php on line 70

KO TwigBundle
```

fabpot commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base.html.twig:
```diff
@@ -3,21 +3,16 @@
-        </style>
     </head>
-    <body>
+    <body id="{% block body_id %}{{ app.request.query.get('panel', '')|lower }}{% endblock %}">
```
We cannot use `app.*` in the web profiler to keep compatibility with Silex.

fabpot commented:
@javiereguiluz Can you test this redesign with Silex 1.x and 2.x, to validate that we do not break it?

javiereguiluz commented:
@fabpot absolutely! I'm going to do it right now.

javiereguiluz committed:
Tweaked the toolbar for Silex 2

javiereguiluz commented:
In https://github.com/symfony/symfony/commit/a808fa5c291741de4b5dd9deb9df848704ec3476 I've made some changes for Silex 1 and 2. Let's review the changes:

### Silex 2

**Old toolbar**

![silex_2_before](https://cloud.githubusercontent.com/assets/73419/9718212/106c3de4-557a-11e5-9f4f-86eb05bd4744.png)

**New toolbar**

![silex_2_after](https://cloud.githubusercontent.com/assets/73419/9718218/16f74cee-557a-11e5-8401-f90963b27531.png)

**New toolbar detail**

![silex_2_after_detail](https://cloud.githubusercontent.com/assets/73419/9718229/34126f70-557a-11e5-91b3-d62cc89b8b99.png)

All profiler panels are exactly the same as Symfony, except the following changes:

**Config panel**

![symfony_vs_silex_2_config](https://cloud.githubusercontent.com/assets/73419/9718240/5788c6e8-557a-11e5-8c20-ae06f9e78db0.png)

**Logger panel** (there is no channel)

![symfony_vs_silex_2_logger](https://cloud.githubusercontent.com/assets/73419/9718245/61e6ed0e-557a-11e5-9f62-db04c1a793c9.png)

**Events panel** (there is no priority)

![symfony_vs_silex_2_events](https://cloud.githubusercontent.com/assets/73419/9718248/68e59146-557a-11e5-95fa-5be7206152f6.png)

### Silex 1

First, I cannot make the profiler work for a freshly installed Silex 1 app because of [this issue](https://github.com/silexphp/Silex-WebProfiler/issues/67). The solution is to add the following in `src/app.php`:

```php
$app->register(new Silex\Provider\HttpFragmentServiceProvider());
```

The new toolbar looks the same, but it suffers the same shortcomings of the old toolbar:

1) Instead of "Silex 2", you see the Symfony version:

![silex_1_vs_2_toolbar](https://cloud.githubusercontent.com/assets/73419/9718282/bd1ddb60-557a-11e5-9240-d5b9fdfba529.png)

2) The configuration panel also cannot display the Silex version:

![silex_1_vs_2_config](https://cloud.githubusercontent.com/assets/73419/9718307/ee04c54a-557a-11e5-896c-50da5cf1d37c.png)

3) The log messages are slightly different:

![silex_1_vs_2_logger](https://cloud.githubusercontent.com/assets/73419/9718310/f5dd9d0a-557a-11e5-8807-699bfed02cfb.png)

hason commented:
Perfect!
I would change some things:

* add responsive design for smaller devices (reintroduce minimizing of menu and remove min-width),
* add support to disable clickable links (show all logs, all events etc. - old behaviour),
* enable the full width on a bigger screen (remove max-width).

javiereguiluz committed:
Removed the app.request... call to make it compatible with Silex

javiereguiluz commented:
@hason I'd prefer to have this PR merged as it is and then let people experience the new profiler before changing any of the things you mentioned :)

fabpot commented:
Thank you @javiereguiluz.
